src/EventListener/UpdateInDoctrineListener.php

```diff
@@ -0,0 +1,70 @@
+<?php
+
+declare(strict_types=1);
+
+namespace KaroIO\MessengerMonitorBundle\EventListener;
+
+use KaroIO\MessengerMonitorBundle\Stamp\MonitorIdStamp;
+use KaroIO\MessengerMonitorBundle\Storage\StoredMessage;
+use KaroIO\MessengerMonitorBundle\Storage\StoredMessageRepository;
+use Symfony\Component\EventDispatcher\EventSubscriberInterface;
+use Symfony\Component\Messenger\Envelope;
+use Symfony\Component\Messenger\Event\WorkerMessageHandledEvent;
+use Symfony\Component\Messenger\Event\WorkerMessageReceivedEvent;
+
+/**
+ * todo: see how retries fit into this.
+ *
+ * @internal
+ */
+final class UpdateInDoctrineListener implements EventSubscriberInterface
+{
+    private $storedMessageRepository;
+
+    public function __construct(StoredMessageRepository $storedMessageRepository)
+    {
+        $this->storedMessageRepository = $storedMessageRepository;
+    }
+
+    public function onMessageReceived(WorkerMessageReceivedEvent $event): void
+    {
+        $storedMessage = $this->getStoredMessage($event->getEnvelope());
+
+        $storedMessage->setReceivedAt(\DateTimeImmutable::createFromFormat('U', (string) time()));
+        $this->storedMessageRepository->updateMessage($storedMessage);
+    }
+
+    public function onMessageHandled(WorkerMessageHandledEvent $event): void
+    {
+        $storedMessage = $this->getStoredMessage($event->getEnvelope());
+
+        $storedMessage->setHandledAt(\DateTimeImmutable::createFromFormat('U', (string) time()));
+        $this->storedMessageRepository->updateMessage($storedMessage);
+    }
+
+    private function getStoredMessage(Envelope $envelope): StoredMessage
+    {
+        /** @var MonitorIdStamp $monitorIdStamp */
+        $monitorIdStamp = $envelope->last(MonitorIdStamp::class);
+
+        if (null === $monitorIdStamp) {
+            throw new \RuntimeException('Envelope should have a MonitorIdStamp!');
+        }
+
+        $storedMessage = $this->storedMessageRepository->findMessage($monitorIdStamp->getId());
+
+        if (null === $storedMessage) {
+            throw new \RuntimeException(sprintf('Message with id "%s" not found', $monitorIdStamp->getId()));
+        }
```

Comment:
These failures will probably need to be "softer" - like not an exception. I don't like "hiding" errors, but if something goes wrong with monitoring, it shouldn't cause the whole message handing system to fail. One real example is if a message is sent to AMQP from an external system. In that situation (unless the user adds it manually in their custom serializer), the message will not have a `MonitorIdStamp`

So, what should we do in this case? I'm not sure. Options:

A) just skip doing anything with the message - no tracking for it
B) add code to guarantee that every message has a `MessageIdStamp`. I'm not sure if that's possible... there is `WorkerMessageReceivedEvent` - we could listen to it and if there is no `MonitorIdStamp`, we could add one. But that event (I believe) doesn't allow you to modify the `Envelope`. I can't think of any other hook point (other than a middleware - but then the user would need to add that to their bus manually - that's a "no go") to do this.

So... it might just be that we can't do anything except for log.

Reply:
This is a really good point! i'll take care in the future to these exceptions, keeping in mind that monitoring should not block the messaging system of the main app.

Have you seen any places where this would be problematic?

i'll go for a log as well, because i can't find any good solution. Indeed there is no way to modify the enveloppe from `WorkerMessageReceivedEvent`